apps/emqx_machine/src/emqx_cluster_rpc.erl

```diff
@@ -275,24 +292,91 @@ trans_query(TnxId) ->
-    catch
-        C : E ->
-            ?SLOG(critical, #{msg => "crash to apply MFA", tnx_id => TnxId, mfa => MFA, exception => C, reason => E}),
-            {false, lists:flatten(io_lib:format("TnxId(~p) apply MFA(~p) crash", [TnxId, MFA]))}
+    Result =
+        try
+            Meta0 = #{tnx_id => TnxId, module => M, function => F, args => ?TO_BIN(A)},
+            Res = erlang:apply(M, F, A),
+            Succeed =
+                case Res of
+                    ok ->
+                        OkMeta = Meta0#{msg => <<"succeeded to apply MFA">>, result => Res},
+                        ?SLOG(notice, OkMeta),
+                        {true, OkMeta};
+                    {ok, _} ->
+                        OkMeta1 = Meta0#{msg => <<"succeeded to apply MFA">>, result => ?TO_BIN(Res)},
+                        ?SLOG(notice, OkMeta1),
+                        {true, OkMeta1};
+                    _ ->
+                        NotOkMeta = Meta0#{msg => <<"failed to apply MFA">>, result => ?TO_BIN(Res)},
+                        ?SLOG(error, NotOkMeta),
+                        {false, NotOkMeta}
+                end,
+            {Succeed, Res}
+        catch
+            C : E ->
+                CrashMeta = #{msg => <<"crash to apply MFA">>, tnx_id => TnxId, exception => C, reason => ?TO_BIN(E),
+                    module => M, function => F, args => ?TO_BIN(A)},
+                ?SLOG(critical, CrashMeta),
+                {{false, CrashMeta}, lists:flatten(io_lib:format("TnxId(~p) apply MFA(~p) crash", [TnxId, MFA]))}
+    end,
+    case Result of
+        {{true, Meta}, OkRes} ->
+            emqx_alarm:deactivate(cluster_rpc_apply_failed, Meta),
+            {true, OkRes};
+        {{false, Meta}, NotOkRes} ->
+            emqx_alarm:activate(cluster_rpc_apply_failed, Meta),
+            {false, NotOkRes}
     end.
+
+confirm_commit({ok, TnxId, _Res} = Init, all, _Gap, Timeout) when Timeout =< 0 ->
+    case node_not_commit(TnxId) of
+        ok -> Init;
+        Error -> Error
+    end;
+confirm_commit({ok, TnxId, _Res} = Init, all, Gap, Timeout) ->
+    timer:sleep(Gap),
+    case node_not_commit(TnxId) of
+        ok -> Init;
+        _Error -> confirm_commit(Init, all, Gap, Timeout - Gap)
+    end;
+confirm_commit(Init, 1, _Gap, _Timeout) -> Init;
+confirm_commit({ok, TnxId, _Res} = Init, SucceedNum, _Gap, Timeout) when Timeout =< 0 ->
+    case node_already_commit(TnxId, SucceedNum) of
+        ok -> Init;
+        _Error ->
+            case node_not_commit(TnxId) of
+                ok -> Init; %% All commit but The succeedNum is greater than the length(nodes()).
+                Error -> Error
+            end
+    end;
+confirm_commit({ok, TnxId, _Res} = Init, SucceedNum, Gap, Timeout) ->
+    timer:sleep(Gap),
+    case node_already_commit(TnxId, SucceedNum) of
+        ok -> Init;
+        _Error -> confirm_commit(Init, SucceedNum, Gap, Timeout - Gap)
+    end;
+confirm_commit(Error, _SucceedNum, _Gap, _Timeout) -> Error.
```

Comment:
these function clauses are rather convoluted, please refactor to make it more readable